wordpress-mobile/WordPress-iOS
# Post Scheduled Date Time Zone Conversions

Fixes #12992
Fixes #13080 
Fixes #13132

This adds date formatters which consider the site's time zone instead of relying on device time zone settings. After this change, times shown in the Settings screen will show the time according to the site's time zone.

<details>
  <summary>Example Post (w/ screenshots)</summary>

Here are some screenshots illustrating the same scheduled post (Scheduled for 10:00AM Central Standard Time, 9:00AM in my local Mountain Standard Time) across the web and iOS products after this change:

<img width="827" alt="Screen Shot 2019-12-17 at 8 31 22 AM" src="https://user-images.githubusercontent.com/3250/71010298-de09cc00-20a8-11ea-8100-94a8c5a2375e.png">

<img width="519" alt="Screen Shot 2019-12-17 at 8 31 04 AM" src="https://user-images.githubusercontent.com/3250/71010338-e9f58e00-20a8-11ea-8fae-e37a8a9951b3.png">

<img width="250" alt="Screen Shot 2019-12-17 at 8 31 04 AM" src="https://user-images.githubusercontent.com/3250/71010373-f5e15000-20a8-11ea-823c-349af5d56dd1.png">&nbsp;&nbsp;<img width="250" alt="Screen Shot 2019-12-17 at 8 31 04 AM" src="https://user-images.githubusercontent.com/3250/71010417-042f6c00-20a9-11ea-9fbf-a99e5f169ec9.png">

</details>


Here are the key changes:

- Added `testDisplayDate` method to test whether the display date is real. Verified this by changing the date formatter to use the incorrect time zone and the test was able to catch it.
- Adds `SiteDateFormatters` to pull blog's time zone from `BlogService`
- Adds new Date Formatter for use in `PostSettingsViewController` **this will change the formatting behavior for that cell regardless of the feature flag**
- Sets the time picker to the current time of the day in the site's time zone

## Reviewing Other Platforms

I checked this behavior against the web client behavior and there are a few things I wanted to point out:

- Typically, the Gutenberg editor shows the date in the site's time zone. One exception to this is when the site's time zone is changed _after_ a post is scheduled. In that case, the timestamp from the original time zone is still shown. I believe this is a bug and could lead to some user confusion since the timestamp is neither in the site time zone or the device time zone. iOS handles this appropriately by showing the date in the site's current time zone.

- The list of posts on the web displays the time in the device time zone. This is consistent with the iOS post list as well, although web uses relative date formats so it's a little clearer when the post is _about_ to be published (i.e. "in 5 minutes").

## Testing

- Set the Site Time Zone to a different Time Zone than the device
- Schedule a post
- Check that the displayed scheduled time matches the site's time zone and is published on that date

## Future Considerations

- The new `SiteDateFormatters` struct could be augmented to automatically cache the date formatters. These settings screens don't have a lot of performance concerns, but caching the formatters might be desirable in the future. The `ActivityDateFormatting` struct is very similar and a more unified and updated date approach
- Gutenberg uses the date and time format set in site settings. I think it would be valuable to use this on mobile somewhere, but it might not fit in the cell. It should be possible to convert the [PHP date format strings](https://www.php.net/manual/en/function.date.php) to [Unicode date format strings](https://www.unicode.org/reports/tr35/tr35-dates.html#Date_Format_Patterns). This would probably be a bigger task to consider for a larger date overhaul.

PR submission checklist:


- [x] ~I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.~ This is a smaller bug fix for a feature that was not released to production.
